File: 2021/15/15.ts
import {
  $,
  and,
  arrEqual,
  every,
  filter,
  flatten,
  gte,
  ints,
  join,
  lines,
  map,
  pipe,
  range,
  readInput,
  split
} from '../../common'

import aStar = require('a-star')

type Grid = number[][]
const ceiling: Grid = $(readInput(), lines, map(pipe(split(), ints)))

const findNeighbors =
  (maxX: number, maxY: number) =>
  ([x, y]) =>
    $(
      [
        [x + 1, y],
        [x, y + 1],
        [x - 1, y],
        [x, y - 1]
      ],
      filter(and(every(gte(0)), ([x, y]) => x <= maxX && y <= maxY))
    )

const findPath = (grid: Grid) => {
  const maxX = grid[0].length - 1
  const maxY = grid.length - 1
  return aStar({
    start: [0, 0],
    isEnd: arrEqual([maxX, maxY]),
    neighbor: findNeighbors(maxX, maxY),
    distance: (_, [x, y]) => grid[y][x],
    heuristic: ([x, y]) => maxX - x + (maxY - y),
    hash: join(',')
  }).cost
}

console.log('Part 1:', findPath(ceiling))

const addAndWrap = (n1: number) => (n2: number) => n1 + n2 > 9 ? ((n1 + n2) % 10) + 1 : n1 + n2

const theActualCeilingWhichIsLarger = $(
  range(5),
  map(row =>
    $(
      range(ceiling.length),
      map(y =>
        $(
          range(5),
          map(x => $(ceiling[y], map(addAndWrap(x + row)))),
          flatten()
        )
      )
    )
  ),
  flatten()
)

console.log('Part 2:', findPath(theActualCeilingWhichIsLarger))

File: 2021/15/15.clj
(ns aoc.2021.15.15
  (:require [aoc.astar :refer [astar]]
            [aoc.common :refer [read-input]]
            [clojure.string :as str]))

(defn find-neighbors [max-x max-y]
  (fn [[x, y]]
    (->> [[(inc x), y]
          [x, (inc y)]
          [(dec x), y]
          [x, (dec y)]]
         (filter (fn [[x, y]]
                   (and (>= x 0)
                        (>= y 0)
                        (<= x max-x)
                        (<= y max-y)))))))

(defn find-path [grid]
  (let [max-x (dec (count (first grid)))
        max-y (dec (count grid))]
    (:cost (astar :start [0 0]
                  :is-end #(= [max-x max-y] %)
                  :get-neighbors (find-neighbors max-x max-y)
                  :calculate-cost (fn [_ [x y]] (get-in grid [y x]))
                  :heuristic (fn [[x y]] (+ (- max-x x) (- max-y y)))))))

(let [ceiling (->> (read-input)
                   (mapv (fn [line]
                           (as-> line l
                             (str/split l #"")
                             (mapv parse-long l)))))]
  (println "Part 1:" (find-path ceiling)))